Upgrade Aesara version pin and unpin SciPy upper limit

michaelosthege opened the pull request:
I'm opening the PR so the CI tells me which tests break.

Closes #5446
Closes #5448

michaelosthege committed:
Upgrade Aesara version pin and unpin SciPy upper limit

Closes #5446
Closes #5448

michaelosthege commented on pymc/distributions/multivariate.py:
```diff
@@ -247,8 +247,9 @@ def dist(cls, mu, cov=None, tau=None, chol=None, lower=True, **kwargs):
-            moment = at.full(moment_size, mu)
+            # MvNormal has ndim_supp == 1. Only the last entry is needed.
+            shape = at.concatenate([size, [mu.shape[-1]]])
+            moment = at.full(shape, mu)
```
Not only the `get_moment`, but also the `rng_fn` need to get updated. This applies to at least the `MvNormal`, `MvStudentT`, `Multinomial` and `DirichletMultinomial`.

I'm not familiar enough with these distributions, so @Sayam753 @ricardoV94 @twiecki please check.

michaelosthege commented on pymc/distributions/shape_utils.py:
```diff
@@ -661,7 +662,9 @@ def maybe_resize(
-            rv_out = change_rv_size_fn(rv_var=rv_out, new_size=shape[:-1])
+            # Into added dims 👇 plus implied dims 👇 but excluding support 👇 dims.
+            new_size = (*shape[:-1], *tuple(rv_out.shape[:ndim_actual - ndim_supp]))
+            rv_out = change_rv_size_fn(rv_var=rv_out, new_size=new_size)
```
The `else:` below may no longer be relevant, but we should first add type hints to the signature, because when I tried to un-indent the `ShapeError` it turned out that some of the `int` parameters are actually `Optional[int]`.

michaelosthege replied:
@ricardoV94 it looks like this whole block in `shape_utils.py` is no longer covered by any tests. It is [covered on `main` (line 661)](https://app.codecov.io/gh/pymc-devs/pymc/blob/main/pymc/distributions/shape_utils.py).
Can we remove it, or is there an edge case that could still trigger these branches?

![grafik](https://user-images.githubusercontent.com/5894642/155184948-0eaba1df-8b21-4b37-8410-adbb77635aea.png)

ricardoV94 replied:
>  is there an edge case that could still trigger these branches?

None that I know of

michaelosthege replied:
Looks like I messed up the ShapeWarning. I'll take a look at it in the morning.

michaelosthege commented on pymc/tests/test_distributions_moments.py:
```diff
@@ -161,7 +161,8 @@ def assert_moment_is_expected(model, expected, check_finite_logp=True):
         random_draw = moment
 
-    assert moment.shape == expected.shape == random_draw.shape
+    assert moment.shape == expected.shape
+    assert expected.shape == random_draw.shape
```
I split the `a == b == c` into two asserts to make the errors easier to understand.

michaelosthege commented on pymc/tests/test_distributions_moments.py:
```diff
@@ -1010,8 +1011,8 @@ def test_moyal_moment(mu, sigma, size, expected):
-        (2, rand2d, np.eye(3), 2, np.full((2, 2, 3), rand2d)),
-        (2, rand2d, np.eye(3), (2, 5), np.full((2, 5, 2, 3), rand2d)),
+        (2, rand2d, np.eye(3), 2, np.full((2, 3), rand2d)),
+        (2, rand2d, np.eye(3), (2, 5, 2), np.full((2, 5, 2, 3), rand2d)),
```
These two used to assert the inconsistent interpretation where size was concatenated in front of all other dimensions for multivariate RVs.

michaelosthege commented on pymc/tests/test_shape_handling.py:
```diff
@@ -323,7 +323,7 @@ def test_simultaneous_size_and_dims(self, with_dims_ellipsis):
 
             # Size does not include support dims, so this test must use a dist with support dims.
-            kwargs = dict(name="y", size=2, mu=at.ones((3, 4)), cov=at.eye(4))
+            kwargs = dict(name="y", size=(2, 3), mu=at.ones((3, 4)), cov=at.eye(4))
```
The `mu` creates an implied dimension which is now required to be included in `size`.

michaelosthege reviewed (commented):
I ran only the tests that were red, so maybe I broke something. (We'll see.)

But at least the following tests are still failing, because the RV implementations need to be fixed:
```
FAILED pymc/tests/test_distributions.py::TestMatchesScipy::test_dirichlet_vectorized[2-a2] - ValueError: input operand has more dimensions than allowed by the axis remapping
FAILED pymc/tests/test_distributions.py::TestMatchesScipy::test_dirichlet_vectorized[size1-a2] - ValueError: shape mismatch: objects cannot be broadcast to a single shape
FAILED pymc/tests/test_distributions.py::TestMatchesScipy::test_dirichlet_vectorized[size2-a1] - ValueError: shape mismatch: objects cannot be broadcast to a single shape
FAILED pymc/tests/test_distributions.py::TestMatchesScipy::test_dirichlet_vectorized[size2-a2] - ValueError: shape mismatch: objects cannot be broadcast to a single shape
FAILED pymc/tests/test_distributions_moments.py::test_dirichlet_moment[a2-7-expected2] - TypeError: Cannot convert Type TensorType(float64, (None, None, None)) (of Variable Elemwise{true_div,no_inplace}.0) ...
FAILED pymc/tests/test_distributions_moments.py::test_dirichlet_moment[a3-size3-expected3] - TypeError: Cannot convert Type TensorType(float64, (None, None, None, None)) (of Variable Elemwise{true_div,no_in...
FAILED pymc/tests/test_distributions_moments.py::test_mv_normal_moment[mu7-cov7-size7-expected7] - ValueError: could not broadcast input array from shape (2,2) into shape (4,5,2)
FAILED pymc/tests/test_distributions_moments.py::test_mvstudentt_moment[2-mu5-cov5-2-expected5] - assert (2, 3) == (2, 2, 3)
FAILED pymc/tests/test_distributions_moments.py::test_mvstudentt_moment[2-mu6-cov6-size6-expected6] - ValueError: operands could not be broadcast together with shapes (2,5,2,3) (2,5,2,1,1)
FAILED pymc/tests/test_distributions_moments.py::test_multinomial_moment[p6-n6-2-expected6] - TypeError: Cannot convert Type TensorType(int64, (None, None, None)) (of Variable Elemwise{Cast{int64}}.0) into ...
FAILED pymc/tests/test_distributions_random.py::TestDirichletMultinomial_1D_n_2D_a::test_distribution - assert (1, 2, 4) == (1, 4)
```

For MvNormal and MvStudentT I think this is just the `rng_fn` (and `get_moment`).

-------------------
I gotta jump now, so anybody please feel free to take over.

OriolAbril commented:
You should also check https://pymc--5474.org.readthedocs.build/en/5474/learn/examples/pymc_overview.html#model-specification to make sure not only the installation finishes but that it is actually working and pymc can be imported and used (same url will work as new commits come in)

ricardoV94 commented on pymc/tests/test_shape_handling.py:
```diff
@@ -438,13 +438,12 @@ def test_mvnormal_shape_size_difference(self):
-        ):
+        # The support dimensions are excluded from size. This is a difference w.r.t. NumPy
+        # and from 2021-05 through 2022-02 the behavior here matched NumPy but warned about
+        # potentially unexpected results.
```
I find this comment confusing. The support dimensions being excluded from size is the official Numpy API, see size explanation for multivariate normal: https://numpy.org/doc/stable/reference/random/generated/numpy.random.multivariate_normal.html

I would also remove the from x to y behavior...

michaelosthege replied:
I left it as an explanation for why there's a `pytest.warns(None)` below. But I can take it out

ricardoV94 replied:
Ah... I think I would take it out yes

ricardoV94 committed:
Move stickbreaking tests

These tests were introduced between pre-existing Dirichlet/Multinomial/DirichletMultionmial tests that belong to conceptually related distributions

ricardoV94 committed:
Separate transitive shape checks into separate asserts

ricardoV94 committed:
Adapt to new `size` kwarg interpretation

Closes #5446

ricardoV94 commented on pymc/distributions/multivariate.py:
```diff
@@ -301,18 +301,13 @@ def _shape_from_params(self, dist_params, rep_param_idx=1, param_shapes=None):
 
-        size = tuple(size or ())
+        # Take chi2 draws and add an axis of length 1 to the right for correct broadcasting below
+        chi2_samples = np.sqrt(rng.chisquare(nu, size=size) / nu)[..., None]
```
@Sayam753 can you have a second look if the new Multivariate rng_fn looks correct?

ricardoV94 commented:
Tests are passing, and I doubled check the following suspicious (initially) non-failing distributions:
- [x] MatrixNormal
- [x] KroneckerNormal
- [x] CAR

ricardoV94 committed:
Fix Wishart shape inference when there are batched parameters

ricardoV94 committed:
Update MatrixNormal shape_inference and remove shape and size restrictions

ricardoV94 committed:
Fix KroneckerNormal ndim_supp

michaelosthege committed:
Remove `maybe_resize` helper function and fix some type hints

michaelosthege commented on pymc/util.py:
```diff
@@ -14,7 +14,7 @@
 import functools
 
-from typing import Dict, List, Tuple, Union
+from typing import Dict, List, Tuple, Union, cast
```
This was a missing import I noticed when running `mypy -p pymc`

michaelosthege commented on pymc/distributions/shape_utils.py:
```diff
@@ -565,26 +569,30 @@ def resize_from_observed(
     ----------
-    shape : tuple
+    shape
         A tuple specifying the final shape of a distribution
-    size : tuple
+    size
```
These hints were inaccurate and I didn't want to add the fully expanded meaning of `Optional[WeakShape]` etc.

michaelosthege reviewed (commented):
With the last commit `mypy -p pymc` has zero complaints about `shape_utils.py` :)

ricardoV94 commented on pymc/distributions/distribution.py:
```diff
@@ -353,17 +352,11 @@ def dist(
-        )
+
+        # Replicate dimensions may be prepended via a shape with Ellipsis as the last element:
+        if shape is not None and Ellipsis in shape:
```
do we check anywhere that the ellipsis is in the last position?

michaelosthege replied:
Yes, there are functions further above that do this.
Also that's the difference between `WeakShape` and `StrongShape` so mypy also tells us about errors.
Actually I ran mypy before pushing and it was satisfied.

If only I had run pre-commit the right way. It's still messing with the environment YAML files every time - we have some kind of line ending problem in the hook. So it's always failing for me and sometimes I overlook black problems because of that.

ricardoV94 commented on pymc/distributions/distribution.py:
```diff
@@ -589,18 +582,11 @@ def dist(
+        # Replicate dimensions may be prepended via a shape with Ellipsis as the last element:
+        if shape is not None and Ellipsis in shape:
+            replicate_shape = cast(StrongShape, shape[:-1])
+            graph = cls.change_size(rv_var=graph, new_size=replicate_shape, expand=True)
```
I don't think I've implemented the expand kwarg in censored... but should

```suggestion
            graph = cls.change_size(rv_var=graph, new_size=replicate_shape)
```

ricardoV94 replied:
https://github.com/pymc-devs/pymc/blob/a3bab7ddc0f4612ad039a7d2755f22fa6c8c0a72/pymc/distributions/censored.py#L103

michaelosthege replied:
is it gonna work at all then?

ricardoV94 replied:
right now it's working if not "expanding". We can put a `NotImplementedError` if expand is `True` in `Censored.change_size` for the time being

michaelosthege replied:
You wanna do that commit?
I gotta go run t the supermarket before the vegetables are gone 🙂

ricardoV94 replied:
Fixed

ricardoV94 committed:
Add expand kwarg to Censored.change_size
